src/drivers/amd/renoir/acp_sp_dma.c

```diff
@@ -0,0 +1,416 @@
+		sp_fifo_addr = (uint32_t)(&pscratchregconfig->acp_receive_fifo_buffer);
+		io_reg_write((PU_REGISTER_BASE + ACP_I2S_RX_FIFOADDR), sp_fifo_addr);
+		io_reg_write((PU_REGISTER_BASE + ACP_I2S_RX_FIFOSIZE), (uint32_t)(SP_FIFO_SIZE));
+
+		/* Receive RINGBUFFER Address and size*/
+		sp_buff_addr = config->elem_array.elems[0].dest & DRAM_ADDRESS_MASK;
+		io_reg_write((PU_REGISTER_BASE + ACP_I2S_RX_RINGBUFADDR), sp_buff_addr);
+		io_reg_write((PU_REGISTER_BASE + ACP_I2S_RX_RINGBUFSIZE), sp_buff_size);
+
+		/* Receive DMA transfer size in bytes */
+		io_reg_write((PU_REGISTER_BASE + ACP_I2S_RX_DMA_SIZE), (uint32_t)(ACP_DMA_TRANS_SIZE));
+
+		/* Watermark size for SP receive fifo - Half of SP buffer size*/
+		io_reg_write((PU_REGISTER_BASE + ACP_I2S_RX_INTR_WATERMARK_SIZE), (sp_buff_size/2));
+
+	} else {
+		tr_err(&acp_sp_tr, " ACP: DMA Config channel direcction undefined %d", channel->direction);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int acp_dai_sp_dma_pm_context_restore(struct dma *dma)
+{
+	//TODO
+	return 0;
+}
+
+static int acp_dai_sp_dma_pm_context_store(struct dma *dma)
+{
+	//TODO
+	return 0;
+}
+
+static int acp_dai_sp_dma_copy(struct dma_chan_data *channel, int bytes,
+			  uint32_t flags)
+{
+	struct dma_cb_data next = {
+		.channel = channel,
+		.elem.size = bytes,
+	};
+	notifier_event(channel, NOTIFIER_ID_DMA_COPY,
+			NOTIFIER_TARGET_CORE_LOCAL, &next, sizeof(next));
+	return 0;
+}
+
+static int acp_dai_sp_dma_probe(struct dma *dma)
+{
+	int channel;
+
+	if (dma->chan) {
+		tr_err(&acp_sp_tr, "ACP SP DMA: Repeated probe");
+		return -EEXIST;
+	}
+	tr_dbg(&acp_sp_tr, "SP DMA: probe");
+	dma->chan = rzalloc(SOF_MEM_ZONE_SYS_RUNTIME, 0,
+			    SOF_MEM_CAPS_RAM, dma->plat_data.channels *
+			    sizeof(struct dma_chan_data));
+	if (!dma->chan) {
+		tr_err(&acp_sp_tr, "SP DMA: Probe failure,unable to allocate channel descriptors");
+		return -ENOMEM;
+	}
+	for (channel = 0; channel < dma->plat_data.channels; channel++) {
+		dma->chan[channel].dma = dma;
+		dma->chan[channel].index = channel;
+		dma->chan[channel].status = COMP_STATE_INIT;
+	}
+	atomic_init(&dma->num_channels_busy, 0);
+	return 0;
+}
+
+static int acp_dai_sp_dma_remove(struct dma *dma)
+{
+	if (!dma->chan) {
+		tr_err(&acp_sp_tr, "DMA: remove called without probe,it's a no-op");
+		return 0;
+	}
+	rfree(dma->chan);
+	dma->chan = NULL;
+	return 0;
+}
+
+static int acp_dai_sp_dma_get_data_size(struct dma_chan_data *channel,
+				   uint32_t *avail, uint32_t *free)
+{
+	uint64_t tx_low, curr_tx_pos, tx_high;
+	uint64_t rx_low, curr_rx_pos, rx_high;
+
+	if (channel->direction == DMA_DIR_MEM_TO_DEV) {
+		tx_low = (uint32_t)io_reg_read(PU_REGISTER_BASE +
+				ACP_I2S_TX_LINEARPOSITIONCNTR_LOW);
+		tx_high = (uint32_t)io_reg_read(PU_REGISTER_BASE +
+				ACP_I2S_TX_LINEARPOSITIONCNTR_HIGH);
+		curr_tx_pos = (uint64_t)((tx_high<<32) | tx_low);
+		*free = (curr_tx_pos - prev_tx_pos) > sp_buff_size ? (curr_tx_pos - prev_tx_pos)%sp_buff_size : curr_tx_pos - prev_tx_pos;
```

Comment:
spaces around the % and do we need to check overflows on 64bit reads from HW.

Reply:
Addressed the Review Comment. Done